### PR TITLE
WIP: documentation fixes to reduce the doxygen errors

### PR DESCRIPTION
**Pull request by jtrmal**
Should be fairly straightforward fixes. Still getting the crash of sphinx-doc, but I will follow up on that

### CODE REVIEW

**Comment by danpovey** [comment]
@jtrmal I dont object, merge if you are done.

**Comment by csukuangfj** [comment]
@jtrmal 

I am able to generate C++ API documentation with exhale.

The following are some screenshots
Screen Shot 2020-03-21 at 22 57 52
Screen Shot 2020-03-21 at 22 58 40
Screen Shot 2020-03-21 at 23 04 02

----

I am unable to host the generated HTML files on GitHub since they occupy more 2.8 GB.

```bash
$ ls -lh *.html | wc -l
4800
i$ ls -lh *.html | head -n5
-rw-rw-r-- 1 fangjunkuang fangjunkuang 2.3M Mar 21 20:14 api.html
-rw-rw-r-- 1 fangjunkuang fangjunkuang 582K Mar 21 20:14 classEbwAmSgmmUpdater.html
-rw-rw-r-- 1 fangjunkuang fangjunkuang 589K Mar 21 20:14 classfst_1_1ArcIterator_3_01GrammarFst_01_4.html
-rw-rw-r-- 1 fangjunkuang fangjunkuang 585K Mar 21 20:14 classfst_1_1ArcIterator_3_01TrivialFactorWeightFst_3_01A_00_01F_01_4_01_4.html
-rw-rw-r-- 1 fangjunkuang fangjunkuang 591K Mar 21 20:14 classfst_1_1ArcticWeightTpl.html
```

There are 4800 HTML files; each occupies about 0.5 MB space.

I am wondering whether disk space will be an issue to replace doxygen with sphinx.

----

BTW: we have to transform `src/doc/*.dox` files to `*.rst` files **manually**. It seems that `exhale`
cannot handle `*.dox` files correctly. 


The conversion process should be straightforward.

-------

I did not use `Doxyfile` since it includes both `*.h` and `*.cc` files; there are some symbols
documented not only in header files but also in source files, which confuses `exhale`
and results in the crash of `sphinx`.

**Comment by csukuangfj** [comment]
I find that PyTorch also uses `Exhale`.

From https://pytorch.org/cppdocs/
> Acknowledgements
This documentation website for the PyTorch C++ universe has been enabled by the **Exhale** project and generous investment of time and effort by its maintainer, svenevs. We thank Stephen for his work and his help with the PyTorch C++ documentation.

Maybe we can use the same theme as PyTorch, both for `C++` and `Python` documentation.

What do you think? @danpovey @jtrmal

**Comment by csukuangfj** [comment]
Here is the `conf.py` for PyTorch

https://github.com/pytorch/pytorch/blob/master/docs/source/conf.py

**Comment by jtrmal** [comment]
I will handle the hosting and stuff -- thanks a lot for your help on this,
I was quite clueless how to proceed
y.

**Comment by jtrmal** [comment]
and yeah, I agree with the acknowledgment. Acknowledgement does not cost
us anything.
y.

**Comment by jtrmal** [comment]
How do you want to proceed with the pybind documentation? Do you want to graft it into this structure or keep it separate?@danpovey? @csukuangfj ?

**Comment by csukuangfj** [comment]
[conf.py.txt](https://github.com/kaldi-asr/kaldi/files/4364451/conf.py.txt)
@jtrmal 

Attached is the `conf.py` that I used to generate the above API documentation.
Notice that `Doxyfile` is not used in `conf.py`.  The generation process is rather long; it
costs at least 2 hours.

------

I like the structure PyTorch uses to organize its documentation (https://github.com/pytorch/pytorch/tree/master/docs)

Documentation related stuff is in `docs` folder. C++ documentation is in `docs/cpp/source`
and Python related is in `docs/source`.

```
├── caffe2
│   ├── Caffe2-with-name-55-tall.png
│   ├── DoxygenLayout-c.xml
│   ├── DoxygenLayout-python.xml
│   ├── DOXYGEN.md
│   ├── footer.html
│   ├── header.html
│   ├── installation.md
│   ├── main.css
│   ├── process.py
│   ├── README.md
│   └── stylesheet.css
├── cpp
│   ├── Makefile
│   ├── requirements.txt
│   └── source
├── libtorch.rst
├── make.bat
├── Makefile
├── requirements.txt
└── source
    ├── autograd.rst
    ├── bottleneck.rst
    ├── checkpoint.rst
    ├── community
```

**Comment by jtrmal** [comment]
Thanks -- did you modify also index.rst? I can compile but see no
content... (first picture)
But when I search for a name, I will get the correct output back (second
picture)
I.e. perhaps there is something wrong with the index.srt?
[image: Screen Shot 2020-03-22 at 5.20.44 PM.png][image: Screen Shot
2020-03-22 at 5.21.45 PM.png]

**Comment by csukuangfj** [comment]
sorry, I forgot to metion that you have to create `index.rst` manually and place it adjacent to `conf.py`.

Attached is the file that I used.

[index.rst.txt](https://github.com/kaldi-asr/kaldi/files/4366454/index.rst.txt)

**Comment by jtrmal** [comment]
I was able to combine pybin11 docs and kaldi autodocs -- see the screenshot
below
What is still missing to figure out is the Docs files from doc/*dox --
somehow these files are converted into an xml, but then I dont see them in
the api/*rst files. Not sure what is the problem.
Any ideas?

[image: Screen Shot 2020-03-23 at 7.03.19 PM.png]

**Comment by csukuangfj** [comment]
@jtrmal  
I am sure if we rewrite `*.dox` to `*.rst` format, they can be rendered correctly.

----

The screenshot you posted is not rendered by GitHub.

**Comment by csukuangfj** [comment]
@jtrmal 
I guess GitHub cannot display images sent as attachments in an E-mail.

**Comment by jtrmal** [comment]
Screen Shot 2020-03-23 at 9 08 56 PM
Screenshot

**Comment by jtrmal** [comment]
Rewriting might work, but the rst's format seems extremely fiddly about spaces and stuff...
Might be quite a lot of work

**Comment by csukuangfj** [comment]
markdown format is also fine.

**Comment by csukuangfj** [comment]
@jtrmal 

The rewriting process is quite **straightforward**.  I feel that `*.rst` is more readable.

Attached is the `doc/about.rst`, which is translated from `doc/about.dox`.

[about.rst.txt](https://github.com/kaldi-asr/kaldi/files/4372649/about.rst.txt)


I have included it in kaldi pybind doc for display purpose only.

The following are screenshots for `about.rst`.
Screen Shot 2020-03-24 at 09 23 53
Screen Shot 2020-03-24 at 09 24 04
Screen Shot 2020-03-24 at 09 24 12
Screen Shot 2020-03-24 at 09 25 46

**Comment by jtrmal** [comment]
The funny thing is... guess what I did while you were working on converting the about.dox... :)
Screen Shot 2020-03-23 at 9 42 10 PM

**Comment by jtrmal** [comment]
I'll checkout this: https://github.com/sourcey/moxygen
I still feel it would be a lot of work, especially for the well-cross-referenced pages

**Comment by csukuangfj** [comment]
@jtrmal 
I have not started to convert `*.dox` to`*.rst` yet. 

`about.rst` is the only file that I converted this morning to illustrate how easy the task is.

Cross-references in `about.rst` are not handled correctly because the links it
referred to have not been converted yet. But cross-referencing in `rst` is **super easy**.
You can refer to current kaldi pybind11 doc for an example.

**Comment by jtrmal** [comment]
No worries, I'm converting the docs. It's not going very quickly, though...

**Comment by jtrmal** [comment]
I converted some of the dox's  -- can anyone have a look and let me know if there is something to be changed? I ignored cross-referencing for now -- I think for that we need the complete set of docs...

File: src/doc/build_setup.rst [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+Where do the compiled binaries go?
+----------------------------------
+
+Currently, the Makefiles do not put the compiled binaries in a special place; they just leave them in the directories where the corresponding code is. Currently, binaries exist in the directories ending with ``bin``, e.g. ``"bin/"``, ``"gmmbin/"``, ``"featbin/"``, ``"fstbin/"``, and ``"lmbin/"`` and so on, all of which are subdirectories of ``"src/"``. In the future, we may designate a single place to put all the binaries.
+
+How our Makefiles work
+----------------------
+
+Currently, the file ``src/Makefil``e just invokes the Makefiles in all the source sub-directories (``src/base``, ``src/matrix`` and so on). These directories have their own Makefiles, all of which share a common structure. They all include the line:
```
Comment (danpovey) [inline src/doc/build_setup.rst]: e in wong place

File: src/doc/chain.dox [hunk 2]
```diff
@@ -48,7 +48,7 @@ namespace kaldi {
      DNN-HMMs (about 5\% relative better), but the system is about 3 times
      faster to decode; training time is probably a bit faster too, but
      we haven't compared it exactly.
-
+   .
```
Comment (danpovey) [inline src/doc/chain.dox]: not sure what the . is for.

File: src/doc/chain.rst [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+The training procedure for 'chain' models
+-----------------------------------------
+The training procedure for chain models is a lattice-free version of MMI, where the denominator state posteriors are obtained by the forward-backward algorithm over an HMM formed from a phone-level decoding graph, and the numerator state posteriors are obtained by a similar forward-backward algorithm but limited to sequences corresponding to the transcript.
+
+For each output index of the neural net (i.e. for each pdf-id), we compute a derivative of the form (numerator occupation probability - denominator occupation probability), and these are propagated back to the network.
+
+The denominator FSTFor the denominator part of the computation we do forward-backward over a HMM. Actually, because we represent it as a finite state acceptor, the labels (pdf-ids) are associated with the arcs and not the states, so it's not really a HMM in the normal formulation, but it's easier think of it as a HMM because we use the forward-backward algorithm to get posteriors. In the code and scripts we refer to it as the 'denominator FST'.
```
Comment (danpovey) [inline src/doc/chain.rst]: FSTFor

**Comment by jtrmal** [comment]
Thanks for the errors!
ad the dox files -- you tell me :) Will you be ok with having doxygen only
for the autodoc and the rest in rst? One of the reason why I 'held' back is
actually that I don't know if you won't object. The other reason is that
the conversion process is effing annoying and tedious.
y.

File: src/doc/about.rst [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+=======================
+
+---------------
+ What is Kaldi?
+---------------
+
+Kaldi is a toolkit for speech recognition written in C++ and licensed under
+the  Apache License v2.0.  Kaldi is intended for use by speech
+recognition researchers.  For more detailed history and list of contributors see
+\ref history.
```
Comment (csukuangfj) [inline src/doc/about.rst]: I am afraid that `*.rst` files converted from `*.dox` files have incorrect syntax of cross-referencing.

File: src/doc/about.rst [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+              month = dec,
+              title = {The Kaldi Speech Recognition Toolkit},
+          booktitle = {IEEE 2011 Workshop on Automatic Speech Recognition and Understanding},
+               year = {2011},
+          publisher = {IEEE Signal Processing Society},
+           location = {Hilton Waikoloa Village, Big Island, Hawaii, US},
+               note = {IEEE Catalog No.: CFP11SRW-USB},
+    }
+
+The paper can be found <a href=http://publications.idiap.ch/downloads/papers/2012/Povey_ASRU2011_2011.pdf> here </a>.
```
Comment (csukuangfj) [inline src/doc/about.rst]: The paper can be found
`here <http://publications.idiap.ch/downloads/papers/2012/Povey_ASRU2011_2011.pdf>`_

<a/> is not valid in rst.

File: src/doc/build_setup.rst [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+The ``"configure"`` script, located in ``src/``, should be run by typing ``./configure``. This script takes various options. For instance, you can run   ``./configure --shared`` if you want to build the setup with shared libraries instead of static libraries (this can make the binaries smaller), and it also takes various options to enable you to specify different math libraries, for instance, to use OpenBlas. Look at the top of the script to see some example command lines. The configure script currently supports are Cygwin, Darwin (which is Apple's version of BSD UNIX), and Linux. The configure script produces a file called ``kaldi.mk``. This file will be textually included by the Makefiles in the subdirectories (see below).
+
+Editing kaldi.mk
+----------------
+
+Changes that you might want to make to kaldi.mk after running "configure" are the following:
+
+*   Changing the debug level:
+
+    -   The default is ``"-O1"``
```
Comment (csukuangfj) [inline src/doc/build_setup.rst]: ``"-O1"``  --> ``-O1``

double quotes can be removed.

**Comment by danpovey** [comment]
@jtrmal I think we can merge this one.

**Comment by jtrmal** [comment]
OK, merging

**Pull request merged**